The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10052589072/job/27784035148) [(plain)](https://github.com/rust-lang-ci/rust/commit/9e82096798a8089c8dd503fadc57459e3d313799/checks/27784035148/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling home v0.5.9
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
---
   Compiling bootstrap v0.0.0 (/Users/runner/work/rust/rust/src/bootstrap)
error[E0599]: no method named `is_success` found for type `bool` in the current scope
##[error]   --> src/core/build_steps/tool.rs:593:18
    |
585 |               let has_changes = !git(Some(&builder.config.src))
586 | |                 .allow_failure()
587 | |                 .run_always()
587 | |                 .run_always()
588 | |                 .args(["diff-index", "--quiet", &commit])
...   |
592 | |                 .run(builder)
593 | |                 .is_success();
    | |_________________|
    |

error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
error[E0599]: no method named `capture_stdout` found for struct `BootstrapCommand` in the current scope
##[error]   --> src/utils/helpers.rs:537:35
    |
537 |     let mut git = git(source_dir).capture_stdout();
    |
   ::: src/utils/exec.rs:59:1
    |
59  | pub struct BootstrapCommand {
```

</details>

--- a/src/bootstrap/src/core/build_steps/tool.rs
+++ b/src/bootstrap/src/core/build_steps/tool.rs
@@ -589,8 +589,7 @@ impl Step for Rustdoc {
                 .arg("--")
                 .arg(librustdoc_src)
                 .arg(rustdoc_src)
-                .run(builder)
-                .is_success();
+                .run(builder);
 
             if !has_changes {
                 let precompiled_rustdoc = builder

--- a/src/bootstrap/src/lib.rs
+++ b/src/bootstrap/src/lib.rs
@@ -968,7 +968,9 @@ impl Build {
         let mut message = String::new();
         let output: CommandOutput = match output {
             // Command has succeeded
-            Ok(output) if output.status.success() => output.into(),
+            Ok(output) if output.status.success() => {
+                CommandOutput::from_output(output, stdout, stderr)
+            }
             // Command has started, but then it failed
             Ok(output) => {
                 writeln!(
@@ -982,7 +984,7 @@ Executed at: {executed_at}"#,
                 )
                 .unwrap();
 
-                let output: CommandOutput = output.into();
+                let output: CommandOutput = CommandOutput::from_output(output, stdout, stderr);
 
                 // If the output mode is OutputMode::Capture, we can now print the output.
                 // If it is OutputMode::Print, then the output has already been printed to

--- a/src/bootstrap/src/utils/exec.rs
+++ b/src/bootstrap/src/utils/exec.rs
@@ -50,7 +50,7 @@ impl OutputMode {
 /// If you want to delay failures until the end of bootstrap, use [delay_failure].
 ///
 /// By default, the command will print its stdout/stderr to stdout/stderr of bootstrap ([OutputMode::Print]).
-/// If you want to handle the output programmatically, use [BootstrapCommand::capture].
+/// If you want to handle the output programmatically, use [BootstrapCommand::run_capture].
 ///
 /// Bootstrap will print a debug log to stdout if the command fails and failure is not allowed.
 ///

--- a/src/bootstrap/src/utils/exec.rs
+++ b/src/bootstrap/src/utils/exec.rs
@@ -223,17 +223,31 @@ pub fn command<S: AsRef<OsStr>>(program: S) -> BootstrapCommand {
 }
 
 /// Represents the output of an executed process.
-#[allow(unused)]
 pub struct CommandOutput {
     status: CommandStatus,
-    stdout: Vec<u8>,
-    stderr: Vec<u8>,
+    stdout: Option<Vec<u8>>,
+    stderr: Option<Vec<u8>>,
 }
 
 impl CommandOutput {
     #[must_use]
     pub fn did_not_start() -> Self {
-        Self { status: CommandStatus::DidNotStart, stdout: vec![], stderr: vec![] }
+        Self { status: CommandStatus::DidNotStart, stdout: None, stderr: None }
+    }
+
+    #[must_use]
+    pub fn from_output(output: Output, stdout: OutputMode, stderr: OutputMode) -> Self {
+        Self {
+            status: CommandStatus::Finished(output.status),
+            stdout: match stdout {
+                OutputMode::Print => None,
+                OutputMode::Capture => Some(output.stdout),
+            },
+            stderr: match stderr {
+                OutputMode::Print => None,
+                OutputMode::Capture => Some(output.stderr),
+            },
+        }
     }
 
     #[must_use]
@@ -259,7 +273,10 @@ impl CommandOutput {
 
     #[must_use]
     pub fn stdout(&self) -> String {
-        String::from_utf8(self.stdout.clone()).expect("Cannot parse process stdout as UTF-8")
+        String::from_utf8(
+            self.stdout.clone().expect("Accessing stdout of a command that did not capture stdout"),
+        )
+        .expect("Cannot parse process stdout as UTF-8")
     }
 
     #[must_use]
@@ -269,7 +286,10 @@ impl CommandOutput {
 
     #[must_use]
     pub fn stderr(&self) -> String {
-        String::from_utf8(self.stderr.clone()).expect("Cannot parse process stderr as UTF-8")
+        String::from_utf8(
+            self.stderr.clone().expect("Accessing stderr of a command that did not capture stderr"),
+        )
+        .expect("Cannot parse process stderr as UTF-8")
     }
 }
 
@@ -277,18 +297,8 @@ impl Default for CommandOutput {
     fn default() -> Self {
         Self {
             status: CommandStatus::Finished(ExitStatus::default()),
-            stdout: vec![],
-            stderr: vec![],
-        }
-    }
-}
-
-impl From<Output> for CommandOutput {
-    fn from(output: Output) -> Self {
-        Self {
-            status: CommandStatus::Finished(output.status),
-            stdout: output.stdout,
-            stderr: output.stderr,
+            stdout: Some(vec![]),
+            stderr: Some(vec![]),
         }
     }
 }

--- a/src/bootstrap/src/utils/helpers.rs
+++ b/src/bootstrap/src/utils/helpers.rs
@@ -534,7 +534,7 @@ pub fn get_closest_merge_base_commit(
     author: &str,
     target_paths: &[PathBuf],
 ) -> Result<String, String> {
-    let mut git = git(source_dir).capture_stdout();
+    let mut git = git(source_dir);
 
     let merge_base = get_git_merge_base(config, source_dir).unwrap_or_else(|_| "HEAD".into());
 

